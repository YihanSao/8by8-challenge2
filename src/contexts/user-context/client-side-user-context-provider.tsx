'use client';
import { useState, type ReactNode } from 'react';
import {
  SendOTPToEmailParams,
  SignUpWithEmailParams,
  SignInWithOTPParams,
  UserContext,
} from './user-context';
import { useRouter } from 'next/navigation';
import type { User } from '@/model/types/user';

/**
 * Props that can be passed from a server component into a
 * {@link ClientSideUserContextProvider} in order to pre-render child components
 * with the user's information pre-populated.
 */
interface ClientSideUserContextProviderProps {
  user: User | null;
  emailForSignIn: string;
  children?: ReactNode;
}

/**
 * Receives a {@link User} and an email address to use for signing in from a
 * server component and provides these as React state variables to child
 * components. Provides methods for signing in, signing out, and more.
 *
 * @param props - {@link ClientSideUserContextProviderProps}
 * @returns A {@link UserContext} provider.
 */
export function ClientSideUserContextProvider(
  props: ClientSideUserContextProviderProps,
) {
  const [user, setUser] = useState<User | null>(props.user);
  const [emailForSignIn, setEmailForSignIn] = useState(props.emailForSignIn);
  const router = useRouter();

  async function signUpWithEmail(params: SignUpWithEmailParams) {
    const response = await fetch('/api/signup-with-email', {
      method: 'POST',
      body: JSON.stringify(params),
    });

    if (!response.ok) {
      throw new Error('Failed to create user.');
    }

    setEmailForSignIn(params.email);

    router.push('/signin-with-otp');
  }

  async function sendOTPToEmail(params: SendOTPToEmailParams) {
    const response = await fetch('/api/send-otp-to-email', {
      method: 'POST',
      body: JSON.stringify(params),
    });

    if (!response.ok) {
      throw new Error('Failed to send one-time passcode');
    }

    setEmailForSignIn(params.email);

    router.push('/signin-with-otp');
  }

  async function resendOTP() {
    const response = await fetch('/api/resend-otp-to-email', {
      method: 'POST',
      body: JSON.stringify({ email: emailForSignIn }),
    });

    if (!response.ok) {
      throw new Error('Failed to send one-time passcode');
    }
  }

  async function signInWithOTP({ otp }: SignInWithOTPParams) {
    const response = await fetch('/api/signin-with-otp', {
      method: 'POST',
      body: JSON.stringify({ email: emailForSignIn, otp }),
    });

    if (!response.ok) {
      throw new Error('Failed to sign in with OTP.');
    }

    const data = await response.json();
    setUser(data.user as User);
  }

  async function signOut() {
    const response = await fetch('/api/signout', {
      method: 'DELETE',
    });

    if (!response.ok) {
      throw new Error('There was a problem signing out.');
    }

    setUser(null);
  }

  /* istanbul ignore next */
  async function restartChallenge() {
    throw new Error('not implemented.');
  }

  return (
    <UserContext.Provider
      value={{
        user,
        emailForSignIn,
        signUpWithEmail,
        sendOTPToEmail,
        resendOTP,
        signInWithOTP,
        signOut,
        restartChallenge,
      }}
    >
      {props.children}
    </UserContext.Provider>
  );
}
