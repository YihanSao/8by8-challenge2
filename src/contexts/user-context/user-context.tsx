'use client';
import { createNamedContext } from '../../hooks/create-named-context';
import type { User } from '../../model/types/user';
import type { Avatar } from '@/model/types/avatar';
import type { UserType } from '@/model/enums/user-type';

interface SignUpWithEmailParams {
  email: string;
  name: string;
  avatar: Avatar;
  type: UserType;
  captchaToken: string;
}

interface SendOTPToEmailParams {
  email: string;
  captchaToken: string;
}

interface SignInWithOTPParams {
  otp: string;
}

interface UserContextType {
  user: User | null;
  emailForSignIn: string;
  signUpWithEmail(params: SignUpWithEmailParams): Promise<void>;
  sendOTPToEmail(params: SendOTPToEmailParams): Promise<void>;
  resendOTP(): Promise<void>;
  signInWithOTP(params: SignInWithOTPParams): Promise<void>;
  signOut(): Promise<void>;
  restartChallenge(): Promise<void>;
}

const UserContext = createNamedContext<UserContextType>('UserContext');

export {
  UserContext,
  type UserContextType,
  type SignUpWithEmailParams,
  type SendOTPToEmailParams,
  type SignInWithOTPParams,
};
