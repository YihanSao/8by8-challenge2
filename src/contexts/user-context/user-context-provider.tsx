import 'server-only';
import { serverContainer } from '@/services/server/container';
import { SERVER_SERVICE_KEYS } from '@/services/server/keys';
import { ClientSideUserContextProvider } from './client-side-user-context-provider';
import type { PropsWithChildren } from 'react';

/**
 * A server component that reads user information from cookies and passes
 * said information into a {@link ClientSideUserContextProvider} in order to
 * pre-populate child components with user information, if the user is signed in.
 */
export async function UserContextProvider({ children }: PropsWithChildren) {
  const auth = serverContainer.get(SERVER_SERVICE_KEYS.Auth);
  const cookies = serverContainer.get(SERVER_SERVICE_KEYS.Cookies);
  const user = await auth.loadSessionUser();
  const emailForSignIn = await cookies.loadEmailForSignIn();

  return (
    <ClientSideUserContextProvider user={user} emailForSignIn={emailForSignIn}>
      {children}
    </ClientSideUserContextProvider>
  );
}
