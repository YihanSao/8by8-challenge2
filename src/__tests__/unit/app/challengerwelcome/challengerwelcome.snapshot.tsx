import ChallengerWelcome from '@/app/challengerwelcome/page';
import { UserContext, type UserContextType } from '@/contexts/user-context';
import { cleanup, render } from '@testing-library/react';

jest.mock('next/navigation', () => ({
  useRouter: () => ({
    prefetch: jest.fn(),
    push: jest.fn(),
  }),
}));

describe('ChallengerWelcome', () => {
  afterEach(cleanup);

  it('renders challenger-welcome page unchanged', () => {
    const { container } = render(
      <UserContext.Provider value={{ user: null } as UserContextType}>
        <ChallengerWelcome />
      </UserContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
