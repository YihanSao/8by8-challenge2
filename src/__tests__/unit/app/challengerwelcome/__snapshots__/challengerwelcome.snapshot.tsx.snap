// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengerWelcome renders challenger-welcome page unchanged 1`] = `
<div>
  <div
    class="outer_container"
  >
    <div
      class="inner_container"
    >
      <section
        class="section_1"
      >
        <img
          alt="background"
          class="background"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="40"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <div
          class="container"
        >
          <img
            alt="8by8 Logo"
            class="logo"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="40"
            src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </div>
      </section>
      <section
        class="section_2"
      >
        <h1
          class="underline"
        >
          Welcome!
        </h1>
        <p
          class="content"
        >
          Closing the voter registration gap has to be a community effort, so we're asking everyone to join us in taking the #8by8Challengeâ€”register 8 friends to register to vote in 8 days!
        </p>
        <button
          class="get_started_btn"
          type="button"
        >
          Get Started
        </button>
        <p
          class="signin_line"
        >
          Already have an account? 
          <a
            href="/signin"
          >
            Sign in
          </a>
        </p>
        <a
          class="teal_link"
          href="/why8by8"
        >
          See why others are doing it
        </a>
      </section>
      <section
        class="content_3"
      >
        <h2
          class="underline"
        >
          Here's How it Works
        </h2>
        <h3
          class="step_header"
        >
          1. Sign Up
        </h3>
        <p
          class="step_text"
        >
          Sign up with your name and email address to get started.
        </p>
        <img
          alt="sign up"
          class="image"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <h3
          class="step_header"
        >
          2. Invite your friends
        </h3>
        <p
          class="step_text"
        >
          Get 8 friends via social media or messaging apps to join your challenge.
        </p>
        <img
          alt="invite your friends"
          class="image"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <h3
          class="step_header"
        >
          3. Friends take action
        </h3>
        <p
          class="step_text"
        >
          Your friends can support your challenge by taking 1 of 3 actions: register to vote, set up election reminders, or take the challenge themselves. You'll earn 1 badge per friend who takes action!
        </p>
        <img
          alt="friends take action"
          class="image"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <h3
          class="step_header"
        >
          4. Win the challenge, get a reward!
        </h3>
        <p
          class="step_text"
        >
          When all 8 of your friends took action in your challenge within 8 days, and you win! Then select and enjoy a reward from one of our amazing partners.
        </p>
        <img
          alt="earn 8 badges in 8 days"
          class="image"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <button
          class="get_started_btn"
          type="button"
        >
          Get Started
        </button>
        <p
          class="signin_line"
        >
          Already have an account?
           
          <a
            class="signin_link_black"
            href="/signin"
          >
            Sign in
          </a>
        </p>
      </section>
    </div>
  </div>
</div>
`;
